### PR TITLE
🐛(search) lazily retrieve es version

### PR DESCRIPTION
**Pull request by jbpenrath**
## Purpose

Retrieve the elasticsearch server version on initialization could creates bug in contexts where elasticsearch host is not accessible. e.g: during a docker build if we run a manage command, an elasticsearch client is initiliazed and build
failed due to a ConnectionError. So a workaround is to lazily retrieve the elasticsearch version the first time we need to know the version then reuse the value.

## Proposal

- [x] Retrieve `ElasticsearchClientCompat7to6.__es_version__` only when it is needed and cache the result

### CODE REVIEW

File: sandbox/settings.py [hunk 1]
```diff
@@ -139,7 +139,7 @@ class Base(StyleguideMixin, DRFMixin, RichieCoursesConfigurationMixin, Configura
 
     # Security
     ALLOWED_HOSTS = []
-    SECRET_KEY = values.Value(None)
+    SECRET_KEY = values.Value("ThisIsAnExampleKeyForDevPurposeOnly")
```
Comment (sampaccoud) [inline sandbox/settings.py:142]: Should be another commit?

File: src/richie/apps/search/elasticsearch.py [hunk 2]
```diff
@@ -24,9 +26,13 @@ def __init__(self, hosts=None, transport_class=Transport, **kwargs):
         of Elasticsearch we're working with.
         """
         super().__init__(hosts=hosts, transport_class=transport_class, **kwargs)
-        # Store the ES version information in memory so we don't have to re-make the
-        # info request constantly.
-        self.__es_version__ = self.info()["version"]["number"][:1]
+
+    @cached_property
+    def __es_version__(self):
```
Comment (sampaccoud) [inline src/richie/apps/search/elasticsearch.py:31]: Cached_property is thread safe so I'm pretty sure this would make the query for each incoming request. You should set it yourself in the client class. Would be nice if a test can evidence that 😏

Comment (jbpenrath) [reply src/richie/apps/search/elasticsearch.py:31]: I think you would say `@cached_property` is not thead safe. I'll try to test this point and update the code accordingly.

Comment (jbpenrath) [reply src/richie/apps/search/elasticsearch.py:31]: I'll make some successive autocomplete request that use `self.__es_version__` and info is not trigger on each request.

File: tests/apps/search/test_elasticsearch_compat_layer.py [hunk 3]
```diff
@@ -22,48 +47,77 @@ class ElasticSearchCompatLayerTestCase(TestCase):
...
+
+                    self.client.get("/api/v1.0/subjets/")
+                    self.client.get("/api/v1.0/levels/")
+                    self.client.get("/api/v1.0/subjets/8312")
+
+                    mock_es_info.assert_called_once()
+                    mock_es_info.reset_mock()
+
+                    # - Autocomplete checks the es_version,
+                    #   info() should not be trigger again
```
Comment (sampaccoud) [inline tests/apps/search/test_elasticsearch_compat_layer.py:117]:
```suggestion
                    #   info() should not be triggered again
```

File: tests/apps/search/test_elasticsearch_compat_layer.py [hunk 4]
```diff
@@ -22,48 +47,77 @@ class ElasticSearchCompatLayerTestCase(TestCase):
...
-        ]
-        bulk_compat(actions=actions, chunk_size=500, client=ES_CLIENT)
-        ES_INDICES_CLIENT.refresh()
+        with mock.patch(
+            "elasticsearch.Elasticsearch.info", wraps=es_client.info
+        ) as mock_es_info:
+            with mock.patch("richie.apps.search.ES_CLIENT", es_client):
+                with mock.patch(
+                    "richie.apps.search.ES_INDICES_CLIENT", es_indices_client
+                ):
```
Comment (sampaccoud) [inline tests/apps/search/test_elasticsearch_compat_layer.py:75]: You can avoid indentation:
```python
with mock.patch("elasticsearch.Elasticsearch.info") as mock_info, mock.patch(
    "richie.apps.search.ES_CLIENT"
) as mock_es_info, mock.patch(
    "richie.apps.search.ES_INDICES_CLIENT"
) as mock_es_indices_client:
    # Perform a bunch of ES actions, from index management all the way to searches
```

Actually you only need to grab `mock_es_info` so:
```python
with mock.patch("elasticsearch.Elasticsearch.info"), mock.patch(
    "richie.apps.search.ES_INDICES_CLIENT"
), mock.patch("richie.apps.search.ES_CLIENT") as mock_es_info:
    # Perform a bunch of ES actions, from index management all the way to searches
```

Not why you mock the two other things but I will trust you on this :stuck_out_tongue:

**Review by sampaccoud — APPROVED** [review]

File: tests/apps/search/test_elasticsearch_compat_layer.py (hunk 4 above)
Comment (jbpenrath) [reply tests/apps/search/test_elasticsearch_compat_layer.py:75]: Because If I do not use a mock here, `ES_CLIENT` and `ES_INDICES_CLIENT` can already be instanciated before and assertion `mock_es_info.assert_called_once()` fails.
